### DISCUSSION TITLE
Support OceanBase dialect

### DISCUSSION THREAD

**Discussion by PaleNeutron** [sqlalchemy/sqlalchemy#11194] [Usage Questions] [answered]
### Describe the use case

[OceanBase](https://en.oceanbase.com/docs/common-oceanbase-database-10000000001166361) is a kind of mysql re-implementation.

But I found it's DDL is slightly different with current sqlalchemy parser.

### Databases / Backends / Drivers targeted

OceanBase, pymysql driver

### constraint

example

```
'  CONSTRAINT `investorportfolio_OBFK_1711354505177325` FOREIGN KEY (`investor_id`) REFERENCES `strategy`.`investor`(`id`) ON UPDATE RESTRICT ON DELETE RESTRICT ,'
```

current pattern

```
        self._re_fk_constraint = _re_compile(
            r"  "
            r"CONSTRAINT +"
            r"%(iq)s(?P<name>(?:%(esc_fq)s|[^%(fq)s])+)%(fq)s +"
            r"FOREIGN KEY +"
            r"\((?P<local>[^\)]+?)\) REFERENCES +"
            r"(?P<table>%(iq)s[^%(fq)s]+%(fq)s"
            r"(?:\.%(iq)s[^%(fq)s]+%(fq)s)?) +"
            r"\((?P<foreign>(?:%(iq)s[^%(fq)s]+%(fq)s(?: *, *)?)+)\)"
            r"(?: +(?P<match>MATCH \w+))?"
            r"(?: +ON DELETE (?P<ondelete>%(on)s))?"
            r"(?: +ON UPDATE (?P<onupdate>%(on)s))?" % kw
        )
```

difference

```
-  r"(?:\.%(iq)s[^%(fq)s]+%(fq)s)?) +"
+ r"(?:\.%(iq)s[^%(fq)s]+%(fq)s)?) *"
```
and allow `ON UPDATE` appear before `ON DELETE`

### KEY

example

```
  KEY `ix_fund_manager_base_info_end_date` (`end_date`) BLOCK_SIZE 16384 LOCAL,
```
current pattern

```
        self._re_key = _re_compile(
            r"  "
            r"(?:(?P<type>\S+) )?KEY"
            r"(?: +%(iq)s(?P<name>(?:%(esc_fq)s|[^%(fq)s])+)%(fq)s)?"
            r"(?: +USING +(?P<using_pre>\S+))?"
            r" +\((?P<columns>.+?)\)"
            r"(?: +USING +(?P<using_post>\S+))?"
            r"(?: +KEY_BLOCK_SIZE *[ =]? *(?P<keyblock>\S+))?"
            r"(?: +WITH PARSER +(?P<parser>\S+))?"
            r"(?: +COMMENT +(?P<comment>(\x27\x27|\x27([^\x27])*?\x27)+))?"
            r"(?: +/\*(?P<version_sql>.+)\*/ *)?"
            r",?$" % quotes
        )
```

difference

```
-   KEY_BLOCK_SIZE
+  (KEY_)?BLOCK_SIZE
```

And add `LOCAL`

### Additional context

I think these changes will not break mysql but if maintainer think MySQL dialect should be kept pure and clean we can create a custom dialect.

Just give me a mini example how to inherit Mysql dialect and use my own `MySQLTableDefinitionParser`, I can't find reference of `MySQLTableDefinitionParser` in sqlalchemy.

**Comment by CaselIT** [comment, marked as answer]
Hi,

> I think these changes will not break mysql but if maintainer think MySQL dialect should be kept pure and clean we can create a custom dialect.

The custom dialect route is definitely the route to follow, since the sqlalchemy maintainers don't currently have the bandwidth to add support for additional dialects. The changes may be small now, but without a test suite any change the the mysql may break this use case.
Also things may differ in other ways too.

Regarding how to create a custom dialect, the easiest solution is to take a look at one of the external ones and adapt as needed. https://docs.sqlalchemy.org/en/20/dialects/index.html#external-dialects There is also a readme tailed for dialect developers https://github.com/sqlalchemy/sqlalchemy/blob/main/README.dialects.rst

Once you have created it feel free to open a PR to add it to that list of dialects.

**Reply by PaleNeutron** [reply]
@CaselIT I create a [dialect](https://github.com/PaleNeutron/sqlalchemy-oceanbase) following the document but keep get warning like:

```
SAWarning: Dialect mysql:pymysql will not make use of SQL compilation caching as it does not set the 'supports_statement_cache' attribute to ``True``.  This can have significant performance implications including some performance degradations in comparison to prior SQLAlchemy versions.  Dialect maintainers should seek to set this attribute to True after appropriate development and testing for SQLAlchemy 1.4 caching support.   Alternatively, this attribute may be set to False which will disable this warning. (Background on this warning at: https://sqlalche.me/e/20/cprf)
```

I'm sure I have set it

https://github.com/PaleNeutron/sqlalchemy-oceanbase/blob/94e51269ed597d0571ea51a291792804bf0fc93f/sqlalchemy_oceanbase/base.py#L18-L22

**Reply by rad-pat** [reply]
I think you're dialect is defined incorrectly, you've defined like oceanbase is a driver. oceanbase should be the dialect and it uses pymysql by default as the driver. Very similar to a dialect for Starrocks that I did some work on, maybe have a look here: https://github.com/StarRocks/starrocks/blob/main/contrib/starrocks-python-client/starrocks/dialect.py

**Reply by CaselIT** [reply]
@PaleNeutron sorry missed this.

That means that there is no attribute `supports_statement_cache` on the dialect class. The easiest thing it to disable the cacheing by setting `supports_statement_cache=False`.
Enabling the cache is also doable for external dialect, but it requires a bit of care if the compilation code changes based on the values of bind parameters. In this case if it's mostly just a subclass of the mysql dialect it may be fine
